### PR TITLE
Enable image distribution on IPFS

### PR DESCRIPTION
**Pull request by ktock**
https://github.com/containerd/nerdctl/issues/465

This commit enables image distribution on IPFS when `ipfs://` prefix is specified before the image name.
(`ipfs daemon` needs running)

**push**: pushes the specified image to IPFS

The following example pushes `ubuntu:20.04` (pulled from DockerHub with `nerdctl pull ubuntu:20.04` in advance) to IPFS

```console
# nerdctl push ipfs://ubuntu:20.04
INFO[0000] pushing image "ubuntu:20.04" to IPFS         
INFO[0000] ensuring image contents                      
INFO[0005] converted "application/vnd.docker.image.rootfs.diff.tar.gzip" to sha256:3645595d053b17b1910effafdcc6d924ff0f915b25ce81a07f76a7b0bb22c964 
bafkreigrkfhzm5bt5xbb7tvrnwzrkvulgipwupc4w24kwudysexmwzyske
```

**pull**: pulls the specified image from IFPS

```console
# nerdctl pull ipfs://bafkreigrkfhzm5bt5xbb7tvrnwzrkvulgipwupc4w24kwudysexmwzyske
bafkreigrkfhzm5bt5xbb7tvrnwzrkvulgipwupc4w24kwudysexmwzyske:                      resolved       |++++++++++++++++++++++++++++++++++++++| 
index-sha256:cfbf2274a7499e2ccf1738793396cce136576786415d4f3da20e500491c21e95:    done           |++++++++++++++++++++++++++++++++++++++| 
manifest-sha256:78ef4cff96127a9c9302508ab8116d692e68f863f43598c00aa202ed216d961c: done           |++++++++++++++++++++++++++++++++++++++| 
config-sha256:92978473846792f54f088079a8be7b26ab384e08271a3c6b2ca699f75c6a0b25:   done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:3645595d053b17b1910effafdcc6d924ff0f915b25ce81a07f76a7b0bb22c964:    done           |++++++++++++++++++++++++++++++++++++++| 
elapsed: 1.2 s                                                                    total:  29.0 M (24.2 MiB/s)                                      
```

**run**:

```console
# nerdctl run --rm -it ipfs://bafkreigrkfhzm5bt5xbb7tvrnwzrkvulgipwupc4w24kwudysexmwzyske echo hello
hello
```

**compose**:

```console
# mkdir /tmp/proj
# cat <<EOF > /tmp/proj/docker-compose.yml
version: "3.4"
services:
  ubuntu:
    image: ipfs://bafkreigrkfhzm5bt5xbb7tvrnwzrkvulgipwupc4w24kwudysexmwzyske
    command: echo hello
EOF
# nerdctl compose up -f /tmp/proj/docker-compose.yml
INFO[0001] Creating container proj_ubuntu_1             
INFO[0000] No non-localhost DNS nameservers are left in resolv.conf. Using default external servers: [nameserver 8.8.8.8 nameserver 8.8.4.4] 
INFO[0000] IPv6 enabled; Adding default IPv6 external servers: [nameserver 2001:4860:4860::8888 nameserver 2001:4860:4860::8844] 
INFO[0001] Attaching to logs                            
ubuntu_1 |hello
INFO[0001] Container "proj_ubuntu_1" exited             
INFO[0001] All the containers have exited               
INFO[0001] Stopping containers (forcibly)               
INFO[0001] Stopping container proj_ubuntu_1             
```

**encryption** (needs to disable estargz conversoin):

```console
# nerdctl pull ghcr.io/stargz-containers/ubuntu:20.04-org
# nerdctl image encrypt --recipient=jwe:/tmp/mypubkey.pem --platform=linux/amd64 ghcr.io/stargz-containers/ubuntu:20.04-org ghcr.io/stargz-containers/ubuntu:20.04-org-enc
sha256:524694948a388c3116478bbbe86c2934a158f51a1d9b6d28459ac9a5fa573e91
# nerdctl push --ipfs-no-estargz ipfs://ghcr.io/stargz-containers/ubuntu:20.04-org-enc
INFO[0000] pushing image "ghcr.io/stargz-containers/ubuntu:20.04-org-enc" to IPFS 
INFO[0000] ensuring image contents                      
bafkreid4rpdokddpdxkjplmbgijyilmdneqzl3vpxzgpsq4764d5rrm4qq
# nerdctl run --rm -it ipfs://bafkreid4rpdokddpdxkjplmbgijyilmdneqzl3vpxzgpsq4764d5rrm4qq echo hello
FATA[0000] failed to extract layer sha256:ccdbb80308cc5ef43b605ac28fac29c6a597f89f5a169bbedbb8dec29c987439: failed to get stream processor for application/vnd.oci.image.layer.v1.tar+gzip+encrypted: exec: "ctd-decoder": executable file not found in $PATH: unknown 
# nerdctl pull --unpack=false ipfs://bafkreid4rpdokddpdxkjplmbgijyilmdneqzl3vpxzgpsq4764d5rrm4qq echo hello
# nerdctl image decrypt --key=/tmp/mykey.pem ipfs://bafkreid4rpdokddpdxkjplmbgijyilmdneqzl3vpxzgpsq4764d5rrm4qq ubuntu:decrypted
sha256:2aa04b1e79ab1ca8af57a5bf67649871cb5fb6118e9cb07660cd6f1f435c5f80
# nerdctl run --rm -it ubuntu:decrypted echo hello
INFO[0000] No non-localhost DNS nameservers are left in resolv.conf. Using default external servers: [nameserver 8.8.8.8 nameserver 8.8.4.4] 
INFO[0000] IPv6 enabled; Adding default IPv6 external servers: [nameserver 2001:4860:4860::8888 nameserver 2001:4860:4860::8844] 
hello
```

~~This commit also adds `nerdctl ipfs push` and `nerdctl ipfs pull` subcommands as the low-level CLI can be used for debugging.~~

Lazy pulling with stagz snapshotter ([doc](https://github.com/containerd/stargz-snapshotter/blob/main/docs/ipfs.md)) is enabled when `--snapshotter=stargz` is specified.

- TODO:
  - agreement on CLI design
  - tests
  - `nerdctl build` should support pulling base images from IPFS  (can be a following-up PR)
    - We can achive this by having a proxy which converts OCI Distribution Spec API to IPFS.
    - blocker: https://github.com/containerd/containerd/pull/6221
      - must be downstreamed to BuildKit. https://github.com/moby/buildkit/pull/2462
  - have an option for image encryption (can be a following-up PR)

### CODE REVIEW

File: cmd/nerdctl/ipfs_pull.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+	})
+
+	// #region platform flags
+	// platform is defined as StringSlice, not StringArray, to allow specifying "--platform=amd64,arm64"
+	ipfsPullCommand.PersistentFlags().StringSlice("platform", nil, "Pull content for a specific platform")
+	ipfsPullCommand.RegisterFlagCompletionFunc("platform", shellCompletePlatforms)
+	ipfsPullCommand.PersistentFlags().Bool("all-platforms", false, "Pull content for all platforms")
+	// #endregion
+
+	ipfsPullCommand.PersistentFlags().String("scheme", "ipfs", "Scheme to pull the CID (\"ipfs\" or \"ipns\")")
```
Comment (AkihiroSuda) [inline cmd/nerdctl/ipfs_pull.go:52]: nit: register flag completion func

File: cmd/nerdctl/ipfs_pull.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	"github.com/containerd/nerdctl/pkg/platformutil"
+	"github.com/containerd/nerdctl/pkg/strutil"
+	httpapi "github.com/ipfs/go-ipfs-http-client"
+
+	"github.com/spf13/cobra"
+)
+
+func newIPFSPullCommand() *cobra.Command {
+	var ipfsPullCommand = &cobra.Command{
+		Use:           "pull",
```
Comment (AkihiroSuda) [inline cmd/nerdctl/ipfs_pull.go:33]: Explain argument

File: cmd/nerdctl/ipfs_push.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+	}
+
+	// #region platform flags
+	// platform is defined as StringSlice, not StringArray, to allow specifying "--platform=amd64,arm64"
+	ipfsPushCommand.PersistentFlags().StringSlice("platform", nil, "Push content for a specific platform")
+	ipfsPushCommand.RegisterFlagCompletionFunc("platform", shellCompletePlatforms)
+	ipfsPushCommand.PersistentFlags().Bool("all-platforms", false, "Push content for all platforms")
+	// #endregion
+
+	ipfsPushCommand.PersistentFlags().Bool("estargz", true, "Convert the image into eStargz")
```
Comment (AkihiroSuda) [inline cmd/nerdctl/ipfs_push.go:54]: If we want `--estargz` flag here, we should add that to the regular `nerdctl image push` too, and deduplicate the code across `nerdctl image push` and `nerdctl ipfs push`. 
(But do we really need `nerdctl ipfs *` commands?)

File: Dockerfile.d/test-integration-rootless.sh [hunk 4]
```diff
@@ -21,5 +21,26 @@ if [[ "$(id -u)" = "0" ]]; then
...
+        else
+            export IPFS_PATH=${HOME}/.ipfs
+            mkdir -p ${IPFS_PATH}
+            nerdctl run -d --name ipfs_host -v ${IPFS_PATH}:/data/ipfs -p 127.0.0.1:5001:5001 \
+                    ghcr.io/stargz-containers/ipfs/go-ipfs:v0.10.0-org  daemon --offline
+            sleep 3
+            echo "ipfs = true" > ${HOME}/.config/containerd-stargz-grpc/config.toml
+            systemctl --user restart stargz-snapshotter.service
+            sleep 3
+        fi
```
Comment (AkihiroSuda) [inline Dockerfile.d/test-integration-rootless.sh:44]: The logic looks too much complicated.

Maybe we should just run IPFS as a systemd service?

Comment (ktock) [reply Dockerfile.d/test-integration-rootless.sh:44]: Rootless nerdctl (inside rootlesskit network namespace) can't access to IPFS daemon API (127.0.0.1:5001) on the host?

Comment (AkihiroSuda) [reply Dockerfile.d/test-integration-rootless.sh:44]: Right. I thought we could just use a public IP.

Comment (AkihiroSuda) [reply Dockerfile.d/test-integration-rootless.sh:44]: Wondering we should have `containerd-rootless-setuptool.sh install-ipfs` to run IPFS daemon in the netns

https://github.com/containerd/nerdctl/blob/5b52e3b2523647c13d41723cb499dbe1b520971a/extras/rootless/containerd-rootless-setuptool.sh#L397-L419

Comment (ktock) [reply Dockerfile.d/test-integration-rootless.sh:44]: SGTM

File: cmd/nerdctl/compose_up_test.go [hunk 5]
```diff
@@ -30,8 +31,51 @@ import (
...
+  wordpress:
+  db:
+`, testutil.WordpressImage, testutil.MariaDBImage))
+}
+
+func TestIPFSComposeUp(t *testing.T) {
+	testutil.DockerIncompatible(t)
+	base := testutil.NewBase(t)
+	requiresStargz(base)
+	base.Env = append(os.Environ(), "CONTAINERD_SNAPSHOTTER=stargz")
```
Comment (AkihiroSuda) [inline cmd/nerdctl/compose_up_test.go:73]: We should test this for overlayfs as well

File: cmd/nerdctl/push.go [hunk 6]
```diff
@@ -47,6 +58,11 @@ func newPushCommand() *cobra.Command {
 	pushCommand.Flags().Bool("all-platforms", false, "Push content for all platforms")
 	// #endregion
 
+	pushCommand.PersistentFlags().Bool("estargz", false, "Convert the image into eStargz (always true in IPFS mode)")
```
Comment (AkihiroSuda) [inline cmd/nerdctl/push.go:61]: Not sure why this has to be always true in IPFS mode.

File: cmd/nerdctl/push.go [hunk 7]
```diff
@@ -47,6 +58,11 @@ func newPushCommand() *cobra.Command {
...
+
+	pushCommand.PersistentFlags().Bool("ipfs-no-estargz", false, "Disable conversion of the image into eStargz in IPFS mode")
+	pushCommand.PersistentFlags().Bool("ipfs-ensure-image", true, "Ensure the IPFS image is fully pulled to containerd")
```
Comment (AkihiroSuda) [inline cmd/nerdctl/push.go:64]: s/pulled/pushed/ (?)

**Comment by ktock** [comment]
CI seems to be cancelled?

```
Error: The operation was canceled.
```

File: go.mod [hunk 8]
```diff
@@ -25,11 +26,16 @@ require (
...
+	github.com/ipfs/go-cid v0.1.0
+	github.com/ipfs/go-ipfs-files v0.0.9
+	github.com/ipfs/go-ipfs-http-client v0.1.0
+	github.com/ipfs/interface-go-ipfs-core v0.5.2
 	github.com/mattn/go-isatty v0.0.14
 	github.com/moby/sys/mount v0.2.0
 	github.com/opencontainers/go-digest v1.0.0
 	github.com/opencontainers/image-spec v1.0.2-0.20211102003311-9a7a9876500e
 	github.com/opencontainers/runtime-spec v1.0.3-0.20211101234015-a3c33d663ebc
+	github.com/pkg/errors v0.9.1
```
Comment (AkihiroSuda) [inline go.mod:38]: Shouldn't be needed

File: docs/ipfs.md [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+```console
+> nerdctl image encrypt --recipient=jwe:mypubkey.pem ubuntu:20.04 ubuntu:20.04-encrypted
+sha256:a5c57411f3d11bb058b584934def0710c6c5b5a4a2d7e9b78f5480ecfc450740
+> nerdctl push ipfs://ubuntu:20.04-encrypted
+INFO[0000] pushing image "ubuntu:20.04-encrypted" to IPFS
+INFO[0000] ensuring image contents
+bafkreifajsysbvhtgd7fdgrfesszexdq6v5zbj5y2jnjfwxdjyqws2s3s4
+```
+
+You can pull the encrypted image from IPFS using `ipfs://` prefix and can decrypt it in the same way as described in [`/docs/ocycrypt.md`](/docs/ocicrypt.md).
```
Comment (AkihiroSuda) [inline docs/ipfs.md:126]: `/docs/` -> `./`

File: docs/ipfs.md [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+```
+
+The pushed image can run on other (IPFS-agnostic) runtimes.
+
+```console
+> docker run --rm -it ghcr.io/ktock/ubuntu:20.04-ipfs echo hello
+hello
+```
+
+:information_source: Note that though the IPFS-enabled image is OCI compatible, some runtimes including [containerd](https://github.com/containerd/containerd/pull/6221) and [podman](https://github.com/containers/image/pull/1403)) had bugs and failed to pull that image. Containerd has already fixed this.
```
Comment (AkihiroSuda) [inline docs/ipfs.md:96]:
```suggestion
:information_source: Note that though the IPFS-enabled image is OCI compatible, some runtimes including [containerd](https://github.com/containerd/containerd/pull/6221) and [podman](https://github.com/containers/image/pull/1403) had bugs and failed to pull that image. Containerd has already fixed this.
```

**Comment by AkihiroSuda** [comment]
Looks good but a couple of nits.
This is awesome :+1:

**Review by AkihiroSuda — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
Probably now we have to increase the CI timeout

https://github.com/containerd/nerdctl/blob/f316402913c4239a2388b7467bcffc9735dd8327/.github/workflows/test.yml#L55

**Comment by ktock** [comment]
> Probably now we have to increase the CI timeout

I'll fix it in this PR.

File: docs/ipfs.md [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+version: "3.8"
+services:
+  ubuntu:
+    image: ipfs://bafkreicq4dg6nkef5ju422ptedcwfz6kcvpvvhuqeykfrwq5krazf3muze
+    command: echo hello
+```
+
+### encryption
+
+You can distribute [encrypted images](./docs/ocicrypt.md) on IPFS using OCIcrypt.
```
Comment (ktock) [inline docs/ipfs.md:112]: Link seems to be broken. I'll fix it in this PR.

Comment (ktock) [reply docs/ipfs.md]: done.

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Pull request merged**